Implement the initial version of invalid transaction proof (take 2)

liuchengxu opened the pull request:
This PR implements the feature of invalid transaction proof creation and verification, the general idea has been described in #1372, please have a look before reviewing. It is not the full story for invalid transaction proof, a few more iterations are necessary to make it complete. Some notable TODOs for the near future:

- Verifiable bundle solution, which requires some refactorings on the existing bundle solution verification in pallet-domains to be reusable. This is probably the next immediate step.
- Verifiable invalid transaction data fetching. This probably relies on the verifiable bundle solution.
- `core-evm` unimplemented. Need to be revisited later.

Supercedes #1404.

Close #1372 

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu committed:
Implement proof generation for InvalidTransactionProof

- Add `check_transaction_validity()` to DomainCoreApi

  This API returns an error in case the given extrinsic is invalid,
  which will potentially include the storage keys to create the
  storage proof for the related states.

- Generate invalid transaction proof when `check_transaction_validity()` returns an error.

  Ship the invalid extrinsic data directly in the proof for now, This also makes
  writing the test easier in this PR. It will be fetched in a verifiable way in the
  next iteration once the bundle solution is verified.

liuchengxu committed:
Add stateful runtime API support for `RuntimeApiLight`

`RuntimeApiLight` is able to work with stateful runtime APIs as long as
the required state is passed as well. This commit adds a method to `RuntimeApiLight`
to support passing the partial state via storage.

This is primarily for the upcoming `InvalidTransactionProofVerifier`, but it's
also useful generally going forward.

liuchengxu committed:
Expose gossip_message_validator in NewFullSystem

It will be used in the following test.

liuchengxu committed:
Introduce InvalidTransactionProofVerifier

This commit implements the invalid transaction proof verification.

liuchengxu committed:
Add `state_root()` in VerifierApi

Used to extract the state from storage proof.

liuchengxu committed:
Integrate InvalidTransactionProofVerifier into service

liuchengxu committed:
Add test for verifying InvalidTransactionProof

- Separate out `create_runtime_api_light()`
- The added test is primarily used for making sure the stateful runtime
  api call works.

liuchengxu committed:
Remove unused code and various minor refactorings

vedhavyas commented on domains/client/domain-executor/src/gossip_message_validator.rs:
```diff
@@ -208,22 +211,57 @@ where
+        let block_number = self
+            .client
+            .number(at)?
+            .ok_or_else(|| sp_blockchain::Error::Backend(format!("Header for #{at} not found")))?;
```
`Number for #{at} not found`

vedhavyas commented on domains/client/domain-executor/src/gossip_message_validator.rs:
```diff
@@ -208,22 +211,57 @@ where
+                .check_transaction_validity(at, extrinsic.clone(), at)?
+            {
+                let storage_keys = match transaction_fee_err {
+                    CheckTxValidityError::Lookup => Default::default(),
```
So its an empty proof for Lookup errors?

liuchengxu replied:
Yes. `check_transaction_validity` returns an error instantly once the lookup fails, no state/storage proof is needed in that case.

NingLin-P replied:
Would be nicer to have some comments on this.

liuchengxu replied:
Refactored to make it self descriptive that the empty storage proof is used for the lookup error.

vedhavyas commented on domains/client/domain-executor/src/gossip_message_validator.rs:
```diff
@@ -208,22 +211,57 @@ where
                 self.parent_chain.submit_fraud_proof_unsigned(fraud_proof)?;
             }
+
+            // TODO: also check invalid XDM?
```
Yes, IIRC it will also have no storage proof for Core domains. We need to provide proof on System domain

liuchengxu replied:
It's a mere placeholder to not forget the invalid XDM, we'll fill in more implementation details once it's decided. It's something we'll discuss deeper along with the other fraud types in the coming weeks.

vedhavyas commented on domains/runtime/core-evm/src/runtime.rs:
```diff
@@ -698,6 +698,13 @@ impl_runtime_apis! {
+            _uxt: <Block as BlockT>::Extrinsic,
+            _block_hash: <Block as BlockT>::Hash,
+        ) -> Result<(), domain_runtime_primitives::CheckTxValidityError> {
+            unimplemented!("TODO: check transaction fee to core-evm")
```
No difference to the other domains. It is same.

liuchengxu replied:
Yes, the only difference is the way of extracting the signer, the other parts should be the same.

vedhavyas replied:
Yeap. Will you be updating this in this PR itself ?

liuchengxu replied:
I planned to add it to this PR but found something more interesting. The existing `extract_signer` implementation in core-evm is inconsistent with the other domains, in which only the extrinsics with valid signatures are returned while those without valid signatures are ignored, the other domains return all the extrisnics as long as it's signed without checking the signature.

https://github.com/subspace/subspace/blob/73fb1c8652ba40ad89391482184536dadcf376f9/domains/runtime/core-evm/src/runtime.rs#L994-L1007

To fix this inconsistency, we need to clarify further whether we need to also check and filter out the invalid transaction before executing the domain block, this backs to our discussion that whether to capture all the potential concrete fraud at the block execution stage. Added a TODO and prefer to have a separate PR to fix all the leftovers in evm domain.

vedhavyas commented on crates/subspace-fraud-proof/src/invalid_transaction_proof.rs:
```diff
@@ -0,0 +1,214 @@
+                .map_err(|_| VerificationError::InvalidStorageProof)
+        };
+
+        let next_fee_multiplier_storage_key = [
```
This is not very pleasing to look at. We create a representation like storage. Here is for Receipts https://github.com/subspace/subspace/blob/main/domains/primitives/messenger/src/messages.rs#L284

liuchengxu replied:
A manual storage representation is actually not the best option here as the storage upstream can possibly be changed in theory, a TODO is added in the last commit to refactor these using a runtime API since they are already known to the runtime.

vedhavyas replied:
We would still need to change them manually in my proopsed one as well. Its easier to know what the structure is when the storage is defined. Raw bytes is always confusing IMO

liuchengxu replied:
resolved TODO in this PR

vedhavyas commented on crates/subspace-fraud-proof/src/invalid_transaction_proof.rs:
```diff
@@ -0,0 +1,217 @@
+        let state_root = self
+            .verifier_client
+            .state_root(*domain_id, *block_number, *domain_block_hash)
+            .expect("Can not fetch state root");
```
better to error out instead of panicking here

liuchengxu replied:
Refactored in later commit.

vedhavyas reviewed (commented):
Looks resonable. Left some nits
I have some concerns over size of the the InvalidTransactionProof once we inlude Bundle and other missing pieces into this since bundle is not part of the runtime.

liuchengxu reviewed (commented):
It's not expected to include the full bundle, the corresponding extrinsic and merkle proof should be fine. But I agree it's something we should consider, the max bundle size is predefined in the domain config, this concern shouldn't persist once it's properly configured.

NingLin-P commented on domains/client/domain-executor/src/core_gossip_message_validator.rs:
```diff
@@ -198,8 +201,13 @@ where
 
-            self.gossip_message_validator
-                .validate_bundle_transactions(&bundle.extrinsics, domain_id)?;
+            let at = bundle_solution.proof_of_election().block_hash;
```
`bundle_solution.proof_of_election().block_hash` refers to the block hash of the system domain:
https://github.com/subspace/subspace/blob/73799ac4879059a8c87ce1c27cad9591c61a6433/crates/sp-domains/src/lib.rs#L214-L217

It is wrong to use this `block_hash` for the core domain.

liuchengxu replied:
Good catch, fixed and refactored the structure to make it less error-prone.

NingLin-P commented on crates/subspace-fraud-proof/src/tests.rs:
```diff
@@ -641,3 +650,138 @@ async fn invalid_execution_proof_should_not_work() {
+        Arc::new(invalid_state_transition_proof_verifier),
+    );
+
+    assert!(
+        proof_verifier
+            .verify(&good_invalid_transaction_proof)
+            .is_ok(),
+        "Valid proof must be accepeted"
+    );
```
`ferdie.produce_blocks(1).await.unwrap();` will give us one more check as the fraud proof will be verified in the block import pipeline.

NingLin-P commented on crates/subspace-fraud-proof/src/tests.rs:
```diff
@@ -641,3 +650,138 @@ async fn invalid_execution_proof_should_not_work() {
+    alice
+        .send_extrinsic(transfer_from_alice_to_one)
+        .await
+        .expect("Failed to send domain extrinsic");
```
We can use `alice.construct_and_send_extrinsic` for convenience.

liuchengxu committed:
Refactor ProofOfElection to reflect it's specific to the system domain

Make it less prone.

Resolves https://github.com/subspace/subspace/pull/1434#discussion_r1186462668

liuchengxu committed:
Minor refactorings to address review comments

liuchengxu committed:
Retrieve the storage keys for verifying tx validity using runtime api

liuchengxu reviewed (commented):
Addressed comments and resolved a few minor TODOs

NingLin-P reviewed (approved):
Make sense overall!

liuchengxu committed:
Merge branch 'main' into impl-invalid-transaction-proof-take-2
